kyo-prelude/shared/src/main/scala/kyo/kernel/Safepoint.scala

```diff
@@ -9,31 +10,64 @@ import scala.util.control.NonFatal
 
 final class Safepoint private () extends Trace.Owner:
 
-    private var depth                            = 0
-    private[kernel] var interceptor: Interceptor = null
-    private val owner                            = Thread.currentThread()
-
-    private def enter(frame: Frame, value: Any): Int =
-        if (Thread.currentThread eq owner) &&
-            depth < maxStackDepth &&
-            (isNull(interceptor) || interceptor.enter(frame, value))
-        then
+    private var state: State             = State.init()
+    private var interceptor: Interceptor = null
+
+    def enter(frame: Frame, value: Any): Boolean =
+        val state    = this.state
+        val threadId = Thread.currentThread().getId()
+        val proceed =
+            state.depth <= maxStackDepth &&
+                state.threadId == threadId &&
+                (!state.hasInterceptor || interceptor.enter(frame, value))
+        if proceed then
+            this.state = state.incrementDepth
             pushFrame(frame)
-            val depth = this.depth
-            this.depth = depth + 1
-            depth + 1
-        else
-            -1
-        end if
+        proceed
     end enter
 
-    private def exit(depth: Int): Unit =
-        this.depth = depth - 1
+    def exit(): Unit =
+        state = state.decrementDepth
+
+    private[kernel] def getInterceptor(): Interceptor = interceptor
+
+    private[kernel] def setInterceptor(newInterceptor: Interceptor): Unit =
+        interceptor = newInterceptor
+        state = state.withInterceptor(newInterceptor != null)
 
 end Safepoint
 
 object Safepoint:
 
+    opaque type State = Long
+
+    object State:
+        // Bit allocation:
+        // Bits 0-15 (16 bits): depth (0-65535)
+        // Bit 16 (1 bit): hasInterceptor flag
+        // Bits 17-63 (47 bits): threadId
+
+        private inline def DepthMask: Long       = 0xffffL
+        private inline def InterceptorMask: Long = 1L << 16
+        private inline def ThreadIdMask: Long    = 0xffffffffffff0000L
+
+        require(maxStackDepth <= 65536)
+
+        inline def init(): State =
+            (Thread.currentThread().getId() << 17) & ThreadIdMask
+
+        extension (state: State)
```

Comment:
should all these be inlined?

Reply:
I tried that but it actually makes the performance worse because Scala generates bytecode with multiple branches:

Not inlined
```java
    public boolean enter(String frame, Object value) {
        boolean proceed;
        long state = this.state;
        long threadId = Thread.currentThread().getId();
        boolean bl = proceed = State$.MODULE$.depth(state) <= 512 && State$.MODULE$.threadId(state) == threadId && (!State$.MODULE$.hasInterceptor(state) || this.kyo$kernel$Safepoint$$interceptor().enter(frame, value));
        if (proceed) {
            this.state = State$.MODULE$.incrementDepth(state);
            Safepoint Owner_this = this;
            int idx = Owner_this.kyo$kernel$Trace$Owner$$inline$index();
            Owner_this.kyo$kernel$Trace$Owner$$inline$frames()[idx & 0xF] = frame;
            Owner_this.kyo$kernel$Trace$Owner$$inline$index_$eq(idx + 1);
        }
        return proceed;
    }
 ```
 
 With inlining:
 ```scala
     public boolean enter(String frame, Object value) {
        Safepoint$ safepoint$;
        long state = this.state;
        long threadId = Thread.currentThread().getId();
        Safepoint$ safepoint$2 = Safepoint$.MODULE$;
        State$ State$_this = State$.MODULE$;
        if ((int)(state & 0xFFFFL) > 512) return false;
        Safepoint$ safepoint$3 = Safepoint$.MODULE$;
        State$ State$_this2 = State$.MODULE$;
        if ((state & 0xFFFFFFFFFFFF0000L) >>> 17 != threadId) return false;
        Safepoint$ safepoint$4 = Safepoint$.MODULE$;
        State$ State$_this3 = State$.MODULE$;
        if ((state & 0x10000L) != 0L) {
            if (!this.kyo$kernel$Safepoint$$interceptor().enter(frame, value)) return false;
        }
        boolean bl = true;
        boolean proceed = bl;
        if (!proceed) return proceed;
        Safepoint$ Safepoint$_this = safepoint$ = Safepoint$.MODULE$;
        this.state = state + 1L;
        Safepoint Owner_this = this;
        int idx = Owner_this.kyo$kernel$Trace$Owner$$inline$index();
        Owner_this.kyo$kernel$Trace$Owner$$inline$frames()[idx & 0xF] = frame;
        Owner_this.kyo$kernel$Trace$Owner$$inline$index_$eq(idx + 1);
        return proceed;
    }
  ```